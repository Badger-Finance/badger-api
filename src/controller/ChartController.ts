import { Controller, Get, PathParams, QueryParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { SettSnapshot } from '../interface/SettSnapshot';
import { SettService } from '../setts/SettsService';

@Controller('/chart/sett')
export class ChartController {
	constructor(private settService: SettService) {}

	@Get('/:settName')
	@ContentType('json')
	getApiLinks(
		@PathParams('settName') settName: string,
		@QueryParams('count') count?: number,
	): Promise<SettSnapshot[]> {
		return this.settService.getSettSnapshots(settName, count);
	}
}
