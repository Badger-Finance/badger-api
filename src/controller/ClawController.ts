import { Controller, Get, PathParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { SponsorData, SyntheticData } from '../interface/Claw';
import { ClawService } from '../service/claw/ClawService';

/*
 * The expiring multiparty (EMP) contract represents a collaterizable synthetic token.
 * We mint these using underlying sett tokens as collateral. In our case its:
 *   - bBADGER -> bCLAW
 *   - bwBTC/ETH SLP -> sCLAW
 *
 * See EMP contract docs for more contract specific info:
 * https://docs-dot-uma-protocol.appspot.com/uma/contracts/ExpiringMultiParty.html
 *
 * See UMA docs for high level info:
 * https://docs.umaproject.org/
 *
 */
@Controller('/claw')
export class ClawController {
  constructor(private clawService: ClawService) {}

  @Get('/emp/:empAddress')
  @ContentType('json')
  async getSyntheticData(@PathParams('empAddress') empAddress: string): Promise<SyntheticData> {
    return await this.clawService.getSyntheticData(empAddress);
  }

  @Get('/emp/:empAddress/sponsor/:sponsorAddress')
  @ContentType('json')
  async getSponsorData(
    @PathParams('empAddress') empAddress: string,
    @PathParams('sponsorAddress') sponsorAddress: string,
  ): Promise<SponsorData> {
    return await this.clawService.getSponsorData(empAddress, sponsorAddress);
  }
}
