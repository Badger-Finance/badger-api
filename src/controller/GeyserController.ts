import { GeyserService } from '../service/geyser/GeyserService';
import { Controller, Get } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { Sett } from '../interface/Sett';

@Controller('/protocol/geyser')
export class GeyserController {
	constructor(private geyserService: GeyserService) {}

	@Get()
	@ContentType('json')
	async listFarms(): Promise<Sett[]> {
		return this.geyserService.listFarms();
	}
}
