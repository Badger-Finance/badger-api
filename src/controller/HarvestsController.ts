import { Controller, Get, QueryParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { HarvestsQueryArgs } from '../args/HarvestsQueryArgs';
import { HarvestsService } from '../service/harvests/HarvestsService';

@Controller('/harvests')
export class HarvestsController {
	constructor(private harvestsService: HarvestsService) {}

	@Get()
	@ContentType('json')
	async listHarvests(@QueryParams() query: HarvestsQueryArgs): Promise<any> {
		return this.harvestsService.listHarvests(query);
	}
}
