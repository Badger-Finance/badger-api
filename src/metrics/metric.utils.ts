import { getAccounts } from '../accounts/accounts.utils';
import { loadChains } from '../chains/chain';
import { Chain } from '../chains/config/chain.config';
import { ProtocolSummary } from '../protocols/interfaces/protocol-summary.interface';
import { getCachedVault } from '../vaults/vaults.utils';
import { ProtocolMetrics, ProtocolSettsMetrics } from './interfaces/metrics.interface';

export const getProtocolMetrics = async (): Promise<ProtocolMetrics> => {
  const [totalUsers, settMetrics] = await Promise.all([getProtocolTotalUsers(), getProtocolSettMetrics()]);
  return { ...settMetrics, totalUsers };
};

export async function getProtocolTotalUsers(): Promise<number> {
  const chains = loadChains();
  const usersAcrossChains = await Promise.all(chains.map((chain) => getAccounts(chain)));
  return new Set([...usersAcrossChains.flat()]).size;
}

export async function getProtocolSettMetrics(): Promise<ProtocolSettsMetrics> {
  const chains = loadChains();
  const multichainsSummary = await Promise.all(chains.map((chain) => getChainMetrics(chain)));

  let totalValueLocked = 0;
  let totalVaults = 0;

  for (const protocolSummary of multichainsSummary) {
    totalValueLocked += protocolSummary.totalValue;
    totalVaults += protocolSummary.setts?.length ?? 0;
  }

  return { totalValueLocked, totalVaults };
}

export async function getChainMetrics(chain: Chain): Promise<ProtocolSummary> {
  const setts = await Promise.all(chain.setts.map((sett) => getCachedVault(sett)));
  const totalValue = setts.reduce((total, sett) => total + sett.value, 0);

  return { totalValue, setts: setts.map(({ name, balance, value }) => ({ name, balance, value })) };
}
