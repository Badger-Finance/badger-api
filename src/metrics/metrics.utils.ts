import { ProtocolMetrics } from '@badger-dao/sdk';
import { NotFound } from '@tsed/exceptions';

import { getAccounts } from '../accounts/accounts.utils';
import { getDataMapper } from '../aws/dynamodb.utils';
import { ProtocolMetricSnapshot } from '../aws/models/protocol-metric-snapshot.model';
import { Chain } from '../chains/config/chain.config';
import { getCachedVault } from '../vaults/vaults.utils';
import { MetricType } from './enums/metric-type';

export async function getChainMetrics(chains: Chain[]): Promise<ProtocolMetrics> {
  let totalUsers = 0;
  let totalValueLocked = 0;
  let totalVaults = 0;

  for (const chain of chains) {
    const accounts = await getAccounts(chain);
    totalUsers += accounts.length;

    const vaults = await chain.vaults.all();
    const vaultSummaries = await Promise.all(
      vaults.map(async (vault) => {
        const { balance, value } = await getCachedVault(chain, vault);
        return { name: vault.name, balance, value };
      }),
    );
    const totalValue = vaultSummaries.reduce((total, vault) => total + vault.value, 0);

    totalValueLocked += totalValue;
    totalVaults += vaultSummaries.length;
  }

  return { totalValueLocked, totalVaults, totalUsers };
}

export async function queryProtocolMetrics(): Promise<ProtocolMetrics> {
  const errorMessage = 'Protocol metrics not available';
  try {
    const mapper = getDataMapper();
    for await (const metric of mapper.query(
      ProtocolMetricSnapshot,
      { type: MetricType.Protocol },
      { scanIndexForward: false, limit: 1 },
    )) {
      return metric;
    }
    throw new NotFound(errorMessage);
  } catch (error) {
    console.error(error);
    throw new NotFound(errorMessage);
  }
}
