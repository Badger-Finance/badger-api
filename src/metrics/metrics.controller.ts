import { Controller, Get, Inject } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';
import { ProtocolMetricModel } from './interfaces/protocol-metric-model';
import { MetricsService } from './metrics.service';

@Controller('/metrics')
export class MetricsController {
  @Inject()
  metricsService!: MetricsService;

  @Get('')
  @ContentType('json')
  @Summary('Get a metric of the protocol across all chains')
  @Description('Returns the total amount of users, total amount of vaults and total value locked across all chains')
  @Returns(200, ProtocolMetricModel)
  @(Returns(404).Description('Protocol metrics not available'))
  async getProtocolMetrics() {
    return this.metricsService.getProtocolMetrics();
  }
}
