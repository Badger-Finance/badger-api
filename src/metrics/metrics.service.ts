import { Service } from '@tsed/common';
import { NotFound } from '@tsed/exceptions';

import { getDataMapper } from '../aws/dynamodb.utils';
import { ProtocolMetricSnapshot } from '../aws/models/protocol-metric-snapshot.model';
import { MetricType } from './enums/metric-type';
import { ProtocolMetrics } from './interfaces/metrics.interface';

@Service()
export class MetricsService {
  async getProtocolMetrics(): Promise<ProtocolMetrics> {
    const mapper = getDataMapper();
    let result: ProtocolMetrics | undefined;

    try {
      for await (const metric of mapper.query(
        ProtocolMetricSnapshot,
        { type: MetricType.protocol },
        { scanIndexForward: false, limit: 1 },
      )) {
        result = metric;
      }
    } catch (error) {
      console.error(error);
      throw new NotFound('Protocol metrics not available');
    }

    if (!result) {
      throw new NotFound('Protocol metrics not available');
    }

    return result;
  }
}
