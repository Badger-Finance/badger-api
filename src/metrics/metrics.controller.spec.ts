import { PlatformServerless } from "@tsed/platform-serverless";
import { PlatformServerlessTest } from "@tsed/platform-serverless-testing";

import { NetworkStatus } from "../errors/enums/network-status.enum";
import { MetricsController } from "./metrics.controller";
import { MetricsService } from "./metrics.service";

describe("MetricsController", () => {
  beforeEach(
    PlatformServerlessTest.bootstrap(PlatformServerless, {
      lambda: [MetricsController]
    })
  );
  afterEach(() => PlatformServerlessTest.reset());

  describe("GET /v2/metrics", () => {
    it("returns metric", async () => {
      jest.spyOn(MetricsService.prototype, "getProtocolMetrics").mockReturnValue(
        Promise.resolve({
          totalUsers: 30_000,
          totalValueLocked: 100_000_000_000,
          totalVaults: 30
        })
      );
      const { body, statusCode } = await PlatformServerlessTest.request.get("/metrics");
      expect(statusCode).toEqual(NetworkStatus.Success);
      expect(JSON.parse(body)).toMatchSnapshot();
    });
  });
});
