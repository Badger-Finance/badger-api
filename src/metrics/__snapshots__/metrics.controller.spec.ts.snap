// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricsController GET /v2/metrics returns metric 1`] = `
Object {
  "totalUsers": 30000,
  "totalValueLocked": 100000000000,
  "totalVaults": 30,
}
`;
